src/dask_awkward/lib/optimize.py

```diff
@@ -86,177 +49,122 @@ def _projectable_io_layer_names(dsk: HighLevelGraph) -> list[str]:
-        try:
-            _attempt_compute_with_columns_brute(dsk, columns=remaining)
-        except IndexError:
-            keep.append(holdout)
-    if keep == columns:
-        keep = None
-    return {pio_layer: keep}
+    """
+    return [n for n, v in dsk.layers.items() if (v.annotations or {}).get("ak_output")]
 
 
 def _has_projectable_awkward_io_layer(dsk: HighLevelGraph) -> bool:
     for _, v in dsk.layers.items():
-        if isinstance(v, AwkwardIOLayer) and hasattr(v.io_func, "project_columns"):
+        if isinstance(v, AwkwardInputLayer) and hasattr(v.io_func, "project_columns"):
             return True
     return False
 
 
-def optimize_iolayer_columns_brute(dsk: HighLevelGraph) -> HighLevelGraph:
-    # if the task graph doesn't contain a column-projectable
-    # AwkwardIOLayer then bail on this optimization (just return the
-    # existing task graph).
+def _mock_output_func(*args: Any, **kwargs: Any) -> Any:
+    import awkward as ak
+
+    for arg in args + tuple(kwargs.values()):
+        if isinstance(arg, ak.Array):
+            arg.layout._touch_data(recursive=True)
+
+
+def _mock_output(layer: Any) -> Any:
+    assert len(layer.dsk) == 1
+
+    new_layer = copy.deepcopy(layer)
+    mp = new_layer.mapping.copy()
+    for k in iter(mp.keys()):
+        mp[k] = (_mock_output_func,) + mp[k][1:]
+    new_layer.mapping = mp
+    return new_layer
+
+
+def _get_column_reports(dsk: HighLevelGraph, keys: Any) -> dict[str, Any]:
     if not _has_projectable_awkward_io_layer(dsk):
-        return dsk
-    # determine the necessary columns to complete the executation of
-    # the metadata (typetracer) based task graph.
-    necessary_cols = _necessary_columns_brute(dsk)
-    layer_name, necessary_cols = list(_necessary_columns_brute(dsk).items())[0]
-
-    if necessary_cols is None:
-        return dsk  # type: ignore
-    layers = dsk.layers.copy()  # type: ignore
-    deps = dsk.dependencies.copy()  # type: ignore
-    new_layer: Layer | None = None
-
-    new_layer = dsk.layers[layer_name].project_columns(necessary_cols)  # type: ignore
-    layers[layer_name] = new_layer
+        return {}
+
+    import awkward as ak
+
+    layers = dsk.layers.copy()
+    deps = dsk.dependencies.copy()
+    reports = {}
+
+    for name in _projectable_input_layer_names(dsk):
+        layers[name], report = layers[name].mock()
+        reports[name] = report
+
+    for name in _output_layer_names(dsk):
+        layers[name] = _mock_output(layers[name])
+
+    hlg = HighLevelGraph(layers, deps)
+    outlayer = list(hlg.layers.values())[-1]
+
+    try:
+        out = get_sync(hlg, list(outlayer.keys())[0])
+    except Exception as err:
+        on_fail = dask.config.get("awkward.optimization.on-fail")
+        if on_fail == "warn":
+            warnings.warn(f"Column projection optimization failed: {type(err)}, {err}")
+            return {}
+        elif on_fail == "pass":
+            return {}
+        else:
+            raise
```

Comment:
@martindurant what do you think about this try-except block?

Reply:
This is reasonable. 

- Should the config parameter be allowed a "-" (as opposed to "_")?
- can we be more specific than Exception? Maybe not.

Reply:
> Should the config parameter be allowed a "-" (as opposed to "_")?

I've tested the config and `-` is OK (dask.array uses dashes), the envvar can be written as `DASK_AWKWARD__OPTIMIZATION__ON_FAIL`

> can we be more specific than Exception? Maybe not.

Yea this crossed my mind as well, as of right now it can be (`AttributeError, ValueError, NotImplementedError, AssertionError`) and catch everything in the current test suite but (a) a list of 4 is kind of big IMO and (b) I'm worried about something else popping up that we don't yet know about.

Reply:
The default is set to `warn` (pinged you in other comment on the diff for the config file)

> If "pass", maybe there should still be a logging.debug call in that block? Maybe add a comment labeling the branch corresponding to the current default.

Sounds good to me

Reply:
> The default is set to warn

OK, then Exception should be fine, and we may look to narrow it down in the future.

Reply:
As with the previous version of optimising, I think it would be useful to allow users a way to run the find-column code on their own dak object without the rest of optimize and without compute. That will help build understanding and should appear in tutorial docs.

Reply:
Yea I've kept the `dask_awkward.lib.inspect.necessary_columns` public function in place. I'll add it to the toplevel `dask_awkward` import